### ISSUE TITLE
Unstable Training Process

### ISSUE THREAD

**Issue by TheTahaaa** [lvwerra/trl#495]
Hello there, 
First of all, I wanna thank you for implementing this fantastic framework.

I'm trying to fine-tune a T5 model for a Seq-to-Seq task using TRL. However, the mean_reward plot just oscillates around the starting point as you can see below.

<img width="1133" alt="Screenshot 2023-07-04 at 2 13 30 PM" src="https://github.com/lvwerra/trl/assets/30401154/647f5e9c-d437-4b0d-97ca-701158eef4ed">
<img width="1324" alt="Screenshot 2023-07-04 at 2 35 24 PM" src="https://github.com/lvwerra/trl/assets/30401154/3f05d7d8-ef1f-4d4d-9412-eeef54805fb4">

I fine-tuned the model on a parallel corpus before using TRL to improve it, and this is the reason why the initial reward signal is relatively high. I've used a learning rate 1e-9 (which is TOO small) for the plot above, since for larger LR like 1e-5, the plot was getting worse than this (It mostly decreased).
I just wanted to ask if is it possible to do something to make the training more stable, or if PPO in this case is not able to improve the model at all. I've already changed hyperparameters like learning_rate, batch_size, mini_batch_size, ppo_epochs and there hasn't been an improvement to the model. Do you think other hyperparameters like vf_coef, clipping_range can make a drastic change in the training process, or I should just accept that PPO doesn't work in this case (since the model has already a good performance before fine-tune by TRL)

I would appreciate any help and comments.

Thanks!

**Comment by lvwerra** [comment]
What's very suspicious is that the KL-div starts off at 40. At the beginning of training when policy and reference model are the same it should be at zero and then slowly increase to maybe 5-10. Find below a plot of the KL evolution in the IMDb example. Have you looked at some generations to see if something off?

<img width="376" alt="Screenshot 2023-07-05 at 10 17 27" src="https://github.com/lvwerra/trl/assets/8264887/a288ebf2-7c26-4110-8d82-91aac2aa0511">

**Comment by TheTahaaa** [comment]
> What's very suspicious is that the KL-div starts off at 40. At the beginning of training when policy and reference model are the same it should be at zero and then slowly increase to maybe 5-10. Find below a plot of the KL evolution in the IMDb example. Have you looked at some generations to see if something off?
> 
> <img alt="Screenshot 2023-07-05 at 10 17 27" width="376" src="https://user-images.githubusercontent.com/8264887/251084281-a288ebf2-7c26-4110-8d82-91aac2aa0511.png">

Hello, and thanks for your response,

I find this quite unexpected as I hadn't realized it before. After making some adjustments, I observed that the KL-divergence now begins at 19 instead of 40. Nevertheless, you are absolutely correct in pointing out that it should ideally start from 0. I have also monitored the generation output, and everything appears to be functioning normally.

It is intriguing to note that the KL-divergence consistently decreases, which goes against my initial expectations. Here is a visualization of the trend:

<img width="354" alt="Screenshot 2023-07-05 at 10 32 51 AM" src="https://github.com/lvwerra/trl/assets/30401154/c4e6ab3d-ca63-4e26-90df-0eff89825207">

**Comment by lvwerra** [comment]
Are the reference model and active model identical at the beginning of training? What adjustments did you make to decrease the KL-div? Also what generation kwargs are you using|?

**Comment by TheTahaaa** [comment]
I did these two adjustments:
1- I set `ref_model = None` in initialising ppo_trainer. 
2- I did override a method as I described [here](https://github.com/lvwerra/trl/issues/470). However, I think this bug is for the case of using PrefixTuning. I'm using LoRA right now for my model (due to better performance), and the bug that I mentioned in that post doesn't happen anymore.  
Also about the generation kwargs, I use simple greedy search and don't use random strategies since the task is seq-to-seq, not a open-end generation. This is my generation_kwargs: generation_kwargs = {"eos_token_id": -1, 'max_new_tokens': 20}

**Comment by lvwerra** [comment]
We had weird results before when not applying sampling, so I would go for the standard generation kwargs for debugging. Maybe checkout this example for T5 and see if you can get it to work on your data: https://github.com/lvwerra/trl/blob/main/examples/sentiment/scripts/t5-sentiment.py

**Comment by TheTahaaa** [comment]
I just use the mentioned generation kwargs, and this time kl-div starts form 14.5. I think there should be another reason for this problem. 
The only thing that I did and I suspect is adding a manual special token at the beginning of the queries (since I use T5 for a seq-to-seq task). However, I don't think it causes the reason because at the end, the query tensor (which the manual special token has been added to it) is given to both models.

**Comment by lvwerra** [comment]
Special tokens should be removed automatically in the `step`. This means that whatever the responses are differ from what the reference model would produce..

**Comment by TheTahaaa** [comment]
Yes, you're right I just see the `step` source code. I will try to debug it and let you know if there was any improvement. 
Thanks for your help!

**Comment by TheTahaaa** [comment]
@lvwerra There is something which grabs my attention in `step()` method source code. As stated [Here](https://github.com/lvwerra/trl/blob/e1531aa52679082e839c3266ac774662046d9942/trl/trainer/ppo_trainer.py#L620), the ref_logprob for a PEFT model is always computed after calling `disable_adapter()` method on the model (it has nothing to do with the `self.ref_model`). I guess `disable_adapter()` causes this difference in kl-div.

**Comment by TheTahaaa** [comment]
@younesbelkada @lvwerra I think this should be considered in computing `ref_log_prob` when the model uses **LoRA**. Am I right?

> There is something which grabs my attention in `step()` method source code. As stated [here](https://github.com/lvwerra/trl/blob/e1531aa52679082e839c3266ac774662046d9942/trl/trainer/ppo_trainer.py#L620), the ref_logprob for a PEFT model is always computed after calling `disable_adapter()` method on the model (it has nothing to do with the `self.ref_model` when the model uses PEFT wrapper). I guess `disable_adapter()` causes this difference in kl-div. I mean, `ppo_step()` compute `ref_log_prob` by disabling the adapters of the active model. However, in case of using an already fine-tuned model **with a LoRA wrapper** as the active model, adapters should be used in computing the `ref_log_prob`.

**Comment by younesbelkada** [comment]
Hi @TheTahaaa 
The `disable_adapter()` method is needed to effectively retrieve the logits of the model with adapters disabled, leading to the same logits as the reference model. 
Can you elaborate more on:

> However, in case of using an already fine-tuned model with a LoRA wrapper as the active model, adapters should be used in computing the ref_log_prob.

Thanks !

**Comment by TheTahaaa** [comment]
Hey @younesbelkada ,
Yes sure!
I might be wrong, but I think when we choose our active model as a fine-tuned model using LoRA, the adapters should be used in computing the logits of the reference model. I guess disable_adapter() disables all the LoRA layers in the model, and it won't lead to the same logits score on the reference model. 
I suspect this because I have seen the value of KL-div wasn't zero in the zero step during my training (as you can see in the first comment of this thread). 
Thanks a lot!

**Comment by younesbelkada** [comment]
Hi @TheTahaaa 
Thanks for providing more details ! 

> when we choose our active model as a fine-tuned model using LoRA, 

Here do you mean having an initial model that has been fine-tuned on your target domain using LoRA, and train new adapters on top of them? Can you share how do you initialize your model before doing PPO?

**Comment by TheTahaaa** [comment]
Sure @younesbelkada .

```
## Load the fine-tuned PEFT model
model_name = 't5-base'
model = AutoModelForSeq2SeqLM.from_pretrained(model_name)
peft_model = PeftModel.from_pretrained(model, "Address to directory")

## Makes the model trainable
for name, param in peft_model.named_parameters():
    if 'lora' in name or 'Lora' in name:
        param.requires_grad = True
peft_model.train()

## Creates TRL model from the peft model
trl_model = trl.AutoModelForSeq2SeqLMWithValueHead.from_pretrained(peft_model)

## Create ppo_trainer
ppo_trainer = PPOTrainer(config, model=trl_model, ref_model=None, tokenizer=tokenizer, dataset=dataset, data_collator=collater)
```

**Comment by laosuan** [comment]
Can I use merge_peft_adapter.py to avoid this issue?

**Comment by TheTahaaa** [comment]
Hey @laosuan, I did override the `step()` method in `PPOTrainer` class, and it worked for me:

<img width="1211" alt="Screenshot 2023-07-11 at 8 35 16 PM" src="https://github.com/lvwerra/trl/assets/30401154/2abf61c1-17ce-4f37-9d74-878ebd5e3417">

Note that I clone the active model before I initialise the ppo_trainer: 
```
ppo_trainer = MyPPOTrainer(ppo_config, 
model=trl_model, ref_model=trl_reference_model, tokenizer=tokenizer, dataset=dataset, data_collator=collator)

```

**Comment by younesbelkada** [comment]
Hi @TheTahaaa 

> I think in the case of using LoRA, we should clone the model and use it as the ref_model (instead of using the active model after disabling the adapters)

I am still puzzled with this statement, if you deactivate the adapters you should get exactly the same logits as the model without adapters, so the reference model. See the figure below:

![image](https://github.com/lvwerra/trl/assets/49240599/c9712779-7220-4abc-b528-6ac9846c073a)

Is it a behavior that you see only on T5 models ? Are you able to reproduce let's say with GPT2?
